### PR TITLE
Make state tipset usage consistent in the API

### PR DESCRIPTION
**Pull request by Stebalien**
_Always_ (almost) use the tipset's parent state, instead of computing.

Exceptions:

* MinerGetBaseInfo. Fixing this would break things so we need to be careful (although we could bump the API version, fix it, then fix the call sites).
* StateReplay. This is replaying a message on top of the given tipset.
* GasEstimateGasLimit. This executes the message on-top-of the tipset's computed state (unlike call which executes it on the tipset's parent state).
  * Having this method and Call apply the message at different heights is really weird.

### CODE REVIEW

**Comment by Stebalien** [comment]
The test failures look problematic.

File: chain/stmgr/utils.go [hunk 1]
```diff
@@ -144,20 +144,6 @@ func MinerSectorInfo(ctx context.Context, sm *StateManager, maddr address.Addres
 	return mas.GetSector(sid)
 }
 
-func GetMinerSectorSet(ctx context.Context, sm *StateManager, ts *types.TipSet, maddr address.Address, snos *bitfield.BitField) ([]*miner.SectorOnChainInfo, error) {
```
Comment (Stebalien) [inline chain/stmgr/utils.go:147]: This was a drive-by fix. Removed because it was only being used in one place at this point.

File: miner/miner.go [hunk 2]
```diff
@@ -510,33 +509,3 @@ func (m *Miner) createBlock(base *MiningBase, addr address.Address, ticket *type
 		WinningPoStProof: wpostProof,
 	})
 }
-
-type actCacheEntry struct {
-	act *types.Actor
-	err error
-}
-
-type cachedActorLookup struct {
```
Comment (Stebalien) [inline miner/miner.go:519]: dead code

File: node/impl/full/state.go [hunk 3]
```diff
@@ -109,7 +108,7 @@ func (a *StateAPI) StateMinerActiveSectors(ctx context.Context, maddr address.Ad
 		return nil, xerrors.Errorf("merge partition active sets: %w", err)
 	}
 
-	return stmgr.GetMinerSectorSet(ctx, a.StateManager, ts, maddr, &activeSectors)
+	return mas.LoadSectors(&activeSectors)
```
Comment (Stebalien) [inline node/impl/full/state.go:111]: This avoids loading the actor twice.

File: node/modules/mpoolnonceapi.go [hunk 4]
```diff
@@ -21,13 +22,41 @@ type MpoolNonceAPI struct {
 	StateAPI full.StateAPI
 }
 
-// GetNonce gets the nonce from actor state
+// GetNonce gets the nonce from current chain head.
 func (a *MpoolNonceAPI) GetNonce(addr address.Address) (uint64, error) {
```
Comment (Stebalien) [inline node/modules/mpoolnonceapi.go:26]: Is this code correct? If so, we may want to do this in the mpool itself to avoid computing tipsets unnecessarily.

**Comment by Stebalien** [comment]
The tests pass but there are still outstanding questions.

File: node/modules/mpoolnonceapi.go (hunk 4 above)
Comment (Stebalien) [reply node/modules/mpoolnonceapi.go:26]: or should there be a +1 here?

**Comment by arajasek** [comment]
1.2.1 is probably the right release for this, after careful review and letting it run on our nodes for a while

File: node/modules/mpoolnonceapi.go [hunk 5]
```diff
@@ -21,13 +22,41 @@ type MpoolNonceAPI struct {
...
+
+	// Otherwise, find the highest nonce in the tipset.
+	msgs, err := a.StateAPI.Chain.MessagesForTipset(ts)
 	if err != nil {
 		return 0, err
 	}
-	return act.Nonce, nil
+	for _, msg := range msgs {
+		vmmsg := msg.VMMessage()
+		if vmmsg.From != keyAddr {
```
Comment (magik6k) [inline node/modules/mpoolnonceapi.go:52]: I think msg.From can technically also be an ID address, though I don't see messages like that on-chain

File: node/modules/mpoolnonceapi.go [hunk 6]
```diff
@@ -21,13 +22,41 @@ type MpoolNonceAPI struct {
...
+			continue
+		}
+		if vmmsg.Nonce > highestNonce {
+			highestNonce = vmmsg.Nonce
```
Comment (magik6k) [inline node/modules/mpoolnonceapi.go:56]: baseActor.Nonce is the next nonce to use (when the actor is created the nonce is 0, first message also has nonce 0, after which actor nonce is set to 1) - so this should be
```suggestion
		if vmmsg.Nonce >= highestNonce {
			highestNonce = vmmsg.Nonce + 1
```

**Review by magik6k — COMMENTED** [review]
Would be nice to finally land this.

Not entirely sure why, but tests fail on latest master with this

File: node/modules/mpoolnonceapi.go (hunk 5 above)
Comment (Stebalien) [reply node/modules/mpoolnonceapi.go:52]: You sure? It looks like message verification will fail if the from address is not a public key.

https://github.com/filecoin-project/lotus/blob/4937fb925e4327831993efbe9c6d47c653f1603e/chain/messagepool/messagepool.go#L635-L637

Comment (Kubuxu) [reply node/modules/mpoolnonceapi.go:52]: Uff we should fix it.

Comment (whyrusleeping) [reply node/modules/mpoolnonceapi.go:52]: yeah, you should abosolutely be able to send a message with an ID address as the from

Comment (Stebalien) [reply node/modules/mpoolnonceapi.go:52]: Well, we can add that to the network upgrade. Note: we'll need to carefully check the VM to make sure we resolve everything as expected before passing it off to the actors.

Comment (Stebalien) [reply node/modules/mpoolnonceapi.go:52]: Note: we currently _implicitly_ resolve the from address when sending. Should we not be doing that?

**Review by magik6k — APPROVED** [review]
Looks good. Also did some manual testing and everything appears to work well
